datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -1138,27 +1138,44 @@ fn numeric_string_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<D
     }
 }
 
+macro_rules! coerce_list_children {
+    ($LHS_TYPE:expr, $RHS_TYPE:expr) => {
+        Arc::new(Arc::unwrap_or_clone(Arc::clone($LHS_TYPE)).with_data_type(
+            comparison_coercion($LHS_TYPE.data_type(), $RHS_TYPE.data_type())?,
+        ))
+    };
+}
+
 /// Coercion rules for list types.
 fn list_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
     use arrow::datatypes::DataType::*;
     match (lhs_type, rhs_type) {
-        (List(_), List(_)) => Some(lhs_type.clone()),
-        (LargeList(_), List(_)) => Some(lhs_type.clone()),
-        (List(_), LargeList(_)) => Some(rhs_type.clone()),
-        (LargeList(_), LargeList(_)) => Some(lhs_type.clone()),
-        (List(_), FixedSizeList(_, _)) => Some(lhs_type.clone()),
-        (FixedSizeList(_, _), List(_)) => Some(rhs_type.clone()),
+        (List(lhs_field), List(rhs_field))
+        | (List(lhs_field), FixedSizeList(rhs_field, _))
+        | (FixedSizeList(lhs_field, _), List(rhs_field)) => {
+            Some(List(coerce_list_children!(lhs_field, rhs_field)))
+        }
+
+        (LargeList(lhs_field), List(rhs_field))
+        | (List(lhs_field), LargeList(rhs_field))
+        | (LargeList(lhs_field), LargeList(rhs_field))
+        | (LargeList(lhs_field), FixedSizeList(rhs_field, _))
+        | (FixedSizeList(lhs_field, _), LargeList(rhs_field)) => {
+            Some(LargeList(coerce_list_children!(lhs_field, rhs_field)))
```

Comment:
nit: handle large list first
nit2 the pattern will be simpler if you put `|`s inside the `(..., ...)`